src/custom/controller.go

```diff
@@ -0,0 +1,175 @@
+		}
+	}
+	return nil
+}
+
+func (cntrlReceiver *controller) mountSecretInDep(
+	// Fetching the latest version of deployment and modifying it
+	ctx context.Context, ns, name string, secret coreV1.Secret, depLabels map[string]string) error {
+	deployment, err := cntrlReceiver.clientSet.AppsV1().Deployments(ns).Get(ctx, name, metaV1.GetOptions{})
+	if err != nil {
+		fmt.Println("Error in cntrlReceiver.clientSet.AppsV1().Deployments(ns).Get(): ", err.Error())
+	}
+	secretVolume := coreV1.Volume{
+		Name: secret.Name + "-secret-volume",
+		VolumeSource: coreV1.VolumeSource{
+			Secret: &coreV1.SecretVolumeSource{
+				SecretName: secret.Name,
+			},
+		},
+	}
+	containerVolumeMount := coreV1.VolumeMount{
+		Name:      secret.Name + "-secret-volume",
+		MountPath: DefaultContainerSecretPath,
+		ReadOnly:  true,
+	}
+	deployment.ObjectMeta = metaV1.ObjectMeta{
+		Name:        name,
+		Namespace:   ns,
+	}
+	// Get and add secret keys from deployment metadata.labels
+	if secretKeys, secretKeysOk := depLabels[DeploymentLabelSecretKeys]; secretKeysOk {
+		secretKeysList := strings.Split(secretKeys, SecretKeysSeparator)
+		for _, key := range secretKeysList {
+			_, keyCheckDataOk := secret.Data[key]
+			_, keyCheckStringDataOk := secret.StringData[key]
+			if !(keyCheckDataOk || keyCheckStringDataOk) {
+				fmt.Printf("Key %s not found in the mentioned secret %s - Skipping mount for key %s\n", key, secret.Name, key)
+				continue
+			}
+			secretVolume.VolumeSource.Secret.Items = append(secretVolume.VolumeSource.Secret.Items, coreV1.KeyToPath{
+				Key:  key,
+				Path: key,
+			})
+		}
+	}
+	// Appending the secret as a volumemount within all the containers in the PodSpec of Deployment
+	deployment.Spec.Template.Spec.Volumes = append(deployment.Spec.Template.Spec.Volumes, secretVolume)
+	for i := range deployment.Spec.Template.Spec.Containers {
+		deployment.Spec.Template.Spec.Containers[i].VolumeMounts = append(
+			deployment.Spec.Template.Spec.Containers[i].VolumeMounts, containerVolumeMount)
+	}
+	// Updating the deployment so that the new secret is present in the newest pods
+	_, err = cntrlReceiver.clientSet.AppsV1().Deployments(ns).Update(ctx, deployment, metaV1.UpdateOptions{})
+	if err != nil {
+		fmt.Println("Error in cntrlReceiver.clientSet.AppsV1().Deployments(ns).Update(): ", err.Error())
+		return err
+	}
+	fmt.Printf("Deployment %s has been updated with desired keys in secret %s\n", name, secret.Name)
+	return nil
+}
+
+func InitController(clientSet kubernetes.Interface, depInformer appsInformers.DeploymentInformer) *controller {
+	newController := &controller{
+		clientSet:      clientSet,
+		depLister:      depInformer.Lister(),
+		depCacheSynced: depInformer.Informer().HasSynced,
+		workQueue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), SecretMounter),
+	}
+	depInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: newController.handleAdd,
+		},
```

Comment:
@shlokchaudhari9 what will happen if we add labels to existing deployment? We are not handling update events?

Reply:
I have updated the code to handle existing deployment updates as well